File: Learnix/backend-node/utils/performanceAnalyzer.js
const admin = require('firebase-admin');
const db = admin.firestore();

// Function to analyze user performance
const analyzePerformance = async (userId) => {
    try {
        const attemptsSnapshot = await db.collection('quizAttempts').where('userId', '==', userId).get();
        const attempts = attemptsSnapshot.docs.map(doc => doc.data());

        if (attempts.length === 0) {
            return { message: 'No attempts found for this user.' };
        }

        const performanceData = {
            totalAttempts: attempts.length,
            averageScore: 0,
            highestScore: 0,
            lowestScore: Infinity,
            skillAnalysis: {}
        };

        attempts.forEach(attempt => {
            performanceData.averageScore += attempt.score;
            performanceData.highestScore = Math.max(performanceData.highestScore, attempt.score);
            performanceData.lowestScore = Math.min(performanceData.lowestScore, attempt.score);

            attempt.skills.forEach(skill => {
                if (!performanceData.skillAnalysis[skill.name]) {
                    performanceData.skillAnalysis[skill.name] = { totalScore: 0, attempts: 0 };
                }
                performanceData.skillAnalysis[skill.name].totalScore += skill.score;
                performanceData.skillAnalysis[skill.name].attempts += 1;
            });
        });

        performanceData.averageScore /= performanceData.totalAttempts;

        // Calculate average score per skill
        for (const skill in performanceData.skillAnalysis) {
            performanceData.skillAnalysis[skill].averageScore = performanceData.skillAnalysis[skill].totalScore / performanceData.skillAnalysis[skill].attempts;
        }

        return performanceData;
    } catch (error) {
        console.error('Error analyzing performance:', error);
        throw new Error('Performance analysis failed.');
    }
};

// Function to record quiz attempt
const recordQuizAttempt = async (attemptData) => {
    try {
        const attemptRef = await db.collection('quizAttempts').add(attemptData);
        return { id: attemptRef.id, ...attemptData };
    } catch (error) {
        console.error('Error recording quiz attempt:', error);
        throw new Error('Failed to record quiz attempt.');
    }
};

// Function to generate performance reports
const generatePerformanceReport = async (userId) => {
    const performanceData = await analyzePerformance(userId);
    // Logic to format and return the report
    return performanceData;
};

// Function to handle AI tutor conversation
const handleTutorConversation = async (userId, message) => {
    // Logic to interact with AI tutor and return response
    return { response: 'AI Tutor response based on the message.' };
};

// Function to analyze quiz attempts and generate analytics
const analyzeQuizAttempts = async () => {
    // Logic to analyze all quiz attempts and generate analytics
};

module.exports = {
    analyzePerformance,
    recordQuizAttempt,
    generatePerformanceReport,
    handleTutorConversation,
    analyzeQuizAttempts
};

File: Learnix/backend-python/ai_services/performance_analyzer.py
from google.cloud import firestore

class PerformanceAnalyzer:
    def __init__(self):
        self.db = firestore.Client()

    def record_quiz_attempt(self, user_id, quiz_id, score, total_questions, attempt_time):
        attempt_data = {
            'user_id': user_id,
            'quiz_id': quiz_id,
            'score': score,
            'total_questions': total_questions,
            'attempt_time': attempt_time,
            'timestamp': firestore.SERVER_TIMESTAMP
        }
        self.db.collection('quiz_attempts').add(attempt_data)

    def analyze_performance(self, user_id):
        attempts_ref = self.db.collection('quiz_attempts').where('user_id', '==', user_id)
        attempts = attempts_ref.stream()

        total_score = 0
        total_attempts = 0
        for attempt in attempts:
            total_score += attempt.to_dict()['score']
            total_attempts += 1

        average_score = total_score / total_attempts if total_attempts > 0 else 0
        return {
            'average_score': average_score,
            'total_attempts': total_attempts
        }

    def generate_report(self, user_id):
        performance_data = self.analyze_performance(user_id)
        report = {
            'user_id': user_id,
            'average_score': performance_data['average_score'],
            'total_attempts': performance_data['total_attempts']
        }
        return report

    def adaptive_quiz_generation(self, user_id):
        performance_data = self.analyze_performance(user_id)
        if performance_data['average_score'] < 50:
            return "Generate easier quiz"
        elif performance_data['average_score'] > 80:
            return "Generate harder quiz"
        else:
            return "Maintain current difficulty level"